ECE-GY 6143 Machine Learning - Spring 2020.    
Name: Sri Sai Hasrshavardhan Reddy Vippala.       
NetID: ssv315

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/drive/14DcdGsfPzzFqkDV2sVy2scd-nApf_z3d?usp=sharing)

***Introduction ***

**Goal Oriented Dialogue Systems Using DeepPavLov**

This project is based on building a goal oriented dialogue system(Conversational AI) for restaurent booking using DeepPavLov framework.The chatbot suggests restaurents based on the type of food the user wants to eat,price range and provides the details about the restaurent.DSTC-2(Dialogue state Tracking Challenge 2)dataset has been used to train the model.The chatbot is built on DeepPavLov which is an open source conversational AI library built on tensorflow and keras.Deepapvlov is used for developing chatbots and complex conversartional systems.We will try to undertand the process involved in collecting information from the user, processing and generating the response by the system.


The project covers the following steps:

0. [Data preparation](#0.-Data-Preparation)
1. [Build Database of items](#1.-Build-Database-of-items)
2. [Build Slot Filler](#2.-Build-Slot-Filler)
3. [Build and Train a Bot](#3.-Build-and-Train-a-Bot)
4. [Interact with bot](#4.-Interact-with-Bot)

Pictures of the developed model deployed in telegram :

![gobot_example.png](https://github.com/deepmipt/DeepPavlov/blob/master/examples/img/gobot_example.png?raw=1)
source:(https://github.com/deepmipt/DeepPavlov/blob/master/examples/img/gobot_example.png?raw=1)

### Overview of the topic

Conversational AI are used extensively in day-to-day life like siri,cortana,alexa,google assistant.Chatbots, which you may have come across while interacting in a food delivery app(uber eats,doordash), salon & restaurent bookings are a type of conversational AI.Conversational chatbots are becoming more common as intensive programming and highly specialized technical skills are no longer needed to create an interactive chatbot.The rise of conversational AI is based on the extensive rise in usage of messaging apps and voice assistant applications.

Chatbots are built on Natural language processing .Chatbots analyse the text,process the information using several complex algorithms to determine what the user meant or what the user's intention is.Based on the intent of the user it determines an appropriate response to the user.

DeepPavLov is a conversational AI framework that conatins all the necessary components to build chatbots.DeepPavLov is used for developing chatbots and complex conversatoional systems.

Machine learning chatbots are characterized with AI,so we need not be precise while taking to the bot.They can understand our language not only our commands and moreover get smarter with time.These models do not have predefined responses.They create the responses from scratch.These models are smarter than traditional bots.They imitate the converstaion of real people.






### Detailed Discussion

![alt text](https://drive.google.com/uc?id=1S_Al54H3M-795wu9Xjnhx84MTYosmQZC)
(figure from http://docs.deeppavlov.ai/en/master/)

A simple bot architecture has three modules: Natural Language understanding(NLU), Dialogue Manager(DM), Natural language generation(NLG).

**Natural Language Understanding(NLU)** receives the user utterence or dialogue.It processes the utterence to identify the domain , intent and entities mentioned by the user.The combination of these three is known as semantic frame.It consists of NER(name-entity recognizer) which helps to build the semantic frame.
 
Utterence : It is the input given by the user. Ex: " Show me today's sports news" is an example of utterence.

Domain : The type of information the user is intending to know about.Ex: In the above diagram the domain of the utterence is weather_service.

Intent : It refers to the itention of the user.For the Ex: "Show me today's sports mews" the intent of the user is to get the news related to sports.Intent usuaaly conatin a verb followed by a noun.In this case "show news" can be used identify the intent of the user.

Entities : They provide more information about the intent of the user.Ex: " Show me todays sports news" the entities are "today" and " sports".These provide metadata about the intent of the user which help to deliver the intent precisely.

The combination of these three is known as semantic frame which is used by the Dialogue manager to generate a response.

**Dialogue Manager(DM)** is responsible for Dialogue state tracking and action policy management.It consists of RNN(Recurrent neural networks)

Dialogue state tracking : It is used to estimate current state of the dialogue."It estimates the be-liefs of possible user’s goals at every dialogue turn.Dialog  state  tracking  is  used  to  estimate  the current belief state of a dialog given all the pre-ceding  conversation" [text link](chrome-extension://oemmndcbldboiebfnladdacbdfmadadm/https://www.aclweb.org/anthology/D18-1299.pdf)

Action policy management : It decides what actions(It may be giving a response to the user or asking for additional information) to be taken after processing the information.

Ex : 
user - " Show me some good restaurents " 

In this case the 
                 
                  domain : restaurents
                  Intent : request_restaurents
                  entities : ( type = "good")
    
Based on this semantic we can have an SQL query to fetch restaurents satisfying the constraint('good').However there are many restaurents satisfying the constraint and we don't want to show the user every restaurent from the DB.So, the policy manager decides what the next action to perform.In this case it require additonal information from the user.DM generates a system action/policy :

                        request_location
so the response from the chatbot will be,

Chatbot - " Can I your Location ? "

**Natural language generation (NLG)** uses generative models which are based on neural networks and rule based templates to convert the system action/policy into natural language to output to the user.

This process runs iteratively till the user is provided with the information.


### Semantic Frame Representation

It requires domain ontology(properties or relations associated with the domain).

speaker 1 : 'find me a cheap Indian restaurent'
             
              'domain' : 'restaurent'
              'slot type' : 'price','type','location'

speaker 2 : 'find me a good movie of leonardo di caprio'

              'domain' : 'movie'
              'slot type' : 'genre','director','year'

The system need s to identify the ontology corresponding to the domain.

Language understanding :

   Domain classification -> Intent Classification -> Slot filling

Given a collection of utterences u(i)  with labels c(i),D = {(u(1),c(1)),.....} where c(i) belongs to C,train a model to estimate labels for new occurances.
So we can apply deep neural net for classification.For ex: we can fit each word.

Explanation is given in between according to the code 








Downloading and installing deepPavLov framework

In [ ]:
!pip install deeppavlov
!python -m deeppavlov install gobot_simple_dstc2

2020-05-17 04:19:51.189 INFO in 'deeppavlov.core.common.file'['file'] at line 32: Interpreting 'gobot_simple_dstc2' as '/usr/local/lib/python3.6/dist-packages/deeppavlov/configs/go_bot/gobot_simple_dstc2.json'


### 0. DATA PREPARATION

DTSC-2 provides actual human conversations in the form of json files.json(javascript object notation) files contain information in a serilaized format and and are used for transmitting structured infromation.
The dialogues are labelled with slots and corresponding dialogue actions which will be used to train dialogue policy network.

The data conatains retaurent reservation  dialogues from DTSC-2.The dataset reader is used for downloading dataset.

In [ ]:

from deeppavlov.dataset_readers.dstc2_reader import SimpleDSTC2DatasetReader

data = SimpleDSTC2DatasetReader().read('my_data')

2020-05-17 04:20:02.216 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from my_data/simple-dstc2-trn.json]
2020-05-17 04:20:02.231 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from my_data/simple-dstc2-val.json]
2020-05-17 04:20:02.316 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from my_data/simple-dstc2-tst.json]
2020-05-17 04:20:02.913 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 282: There are 479 samples in train split.
2020-05-17 04:20:02.915 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 283: There are 6231 samples in valid split.
2020-05-17 04:20:02.915 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 284: There are 6345 samples in test split.


The dataset contains testing,training and validation data for training and testing the model.It also contains templates for forming responses.

In [ ]:
!ls my_data

simple-dstc2-templates.txt  simple-dstc2-trn.json  simple-dstc2-val.json
simple-dstc2-trn.full.json  simple-dstc2-tst.json


The following command can be used to show the templates used by the chatbot to form responses. The word in the templates starting with # are replaced with value from the corresponding database and output as responses from the system.

In [ ]:
!head -n 101 my_data/simple-dstc2-templates.txt

api_call	api_call area="#area" food="#food" pricerange="#pricerange"
bye	You are welcome!
canthear	Sorry, I can't hear you.
canthelp_area	I'm sorry but there is no #area american restaurant in the #area of town.
canthelp_area_food	Sorry there is no #food restaurant in the #area of town.
canthelp_area_food_pricerange	Sorry there is no #pricerange restaurant in the #area of town serving #food food.
canthelp_area_pricerange	Sorry there is no #pricerange restaurant in the #area of town serving #area american food.
canthelp_food	I am sorry but there is no #food restaurant that matches your request.
canthelp_food_pricerange	Sorry there is no #food restaurant in the #pricerange price range.
confirm-domain	You are looking for a restaurant is that right?
expl-conf_area	Did you say you are looking for a restaurant in the #area of town?
expl-conf_food	You are looking for a #food restaurant right?
expl-conf_pricerange	Let me confirm, You are looking for a restaurant in the #pricerange price range 

Similarly we can the data in the training dataset.Data is stored in json format.The data is in the form of slot and corresponding values.

                    "slots": 'area','south'
                    "speaker" : 2

In [ ]:
!head -n 101 my_data/simple-dstc2-trn.json

[
  [
    {
      "speaker": 2,
      "text": "Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?",
      "slots": [],
      "act": "welcomemsg"
    },
    {
      "speaker": 1,
      "text": "cheap restaurant",
      "slots": [
        [
          "pricerange",
          "cheap"
        ]
      ]
    },
    {
      "speaker": 2,
      "text": "What kind of food would you like?",
      "slots": [],
      "act": "request_food"
    },
    {
      "speaker": 1,
      "text": "any",
      "slots": [
        [
          "this",
          "dontcare"
        ]
      ]
    },
    {
      "speaker": 2,
      "text": "What part of town do you have in mind?",
      "slots": [],
      "act": "request_area"
    },
    {
      "speaker": 1,
      "text": "south",
      "slots": [
        [
          "area",
          "south"
        ]
      ]
    },
    {
      "speaker": 2,
      "text": "api_call area=\"south\" food

When we download data, we need to make groups of dialogues referred to as batches.Batches during training phase will be the input to the neural network which will perform weight updations using gradient descent.

The way batches are formed are controlled by dataset iterator.

    "iterator" is our iterator(in this case) which iterates over data batches.

In [ ]:
from deeppavlov.dataset_iterators.dialog_iterator import DialogDatasetIterator

iterator = DialogDatasetIterator(data)

In this model batches are series of two dialogues.

A batch consists of dialogues.A dialogue refers to one utterence from the speaker and one utterence from the chatbot/system.

### Whats in a batch?

the following figure shows a representaion of semantic frames of both speaker and system response

![alt text](https://drive.google.com/uc?id=19RF3AxQdcOqxCduBRqMwzhhRoQL9frSB)
(figure by self)


db_result is the response from the database.

The system response is generated using API calls to the database to retrieve information from relevant Db.
In this case the system makes an API call to the pub_Db which has its attributes

          'pricerange':
          'name':
          'address':
          'phone number':

we can know iterate over batches of preprocessed DTSC-2 dialogues.we can change our batch size to include more responses from the user and system to the model.As we inrease the batch size we have batches of more number of dialogues from both user and the system.






In [ ]:
from pprint import pprint

for dialog in iterator.gen_batches(batch_size=2, data_type='train'):    # In the original notebook batch_size was 1 
    turns_x, turns_y = dialog
    
    print("User utterances:\n----------------\n")
    pprint(turns_x[0], indent=4)
    print("\nSystem responses:\n-----------------\n")
    pprint(turns_y[0], indent=4)
    
    break

User utterances:
----------------

[   {'prev_resp_act': None, 'text': ''},
    {'prev_resp_act': 'welcomemsg', 'text': 'unintelligible'},
    {'prev_resp_act': 'request_area', 'text': 'unintelligible'},
    {   'db_result': {   'addr': '3 - 5 millers yard mill lane',
                         'area': 'centre',
                         'food': 'indian',
                         'name': 'mahal of cambridge',
                         'phone': '01223 360409',
                         'pricerange': 'cheap'},
        'prev_resp_act': 'api_call',
        'text': 'unintelligible'},
    {'prev_resp_act': 'offer_name', 'text': 'unintelligible'},
    {   'prev_resp_act': 'offer_name',
        'slots': [['pricerange', 'moderate']],
        'text': 'moderate'}]

System responses:
-----------------

[   {   'act': 'welcomemsg',
        'text': 'Hello, welcome to the Cambridge restaurant system. You can '
                'ask for restaurants by area, price range or food type. How '
                'm

As training all the dialogues is edxpensive and time-consuming we have limited the training to 50 dialogues

In [ ]:
!cp my_data/simple-dstc2-trn.json my_data/simple-dstc2-trn.full.json

In [ ]:
import json

NUM_TRAIN = 50

with open('my_data/simple-dstc2-trn.full.json', 'rt') as fin:
    data = json.load(fin)
with open('my_data/simple-dstc2-trn.json', 'wt') as fout:
    json.dump(data[:NUM_TRAIN], fout, indent=2)
print(f"Train set is reduced to {NUM_TRAIN} dialogues (out of {len(data)}).")

Train set is reduced to 50 dialogues (out of 50).


## 1. Build Database of items

we need to setup a database so that we can retrieve the information about restaurents whenever the chatbot makes an api call.The database contains task-specific information such as food-type,pricerange,location,phone number etc.

We can see how the input from the user is processed.The sentence is split intlo to identify domain,intent and entities and store them in slots.These slots are used when chatbot makes api calls to database to retrieve relevant information.

Consider the example, Based on the slot values of attributes 'pricerange' and 'area' the chatbot makes an api call to the database to retrieve the information n regarding entries matching the given constraints.Suppose the database contains the following entries corresponding to the given slot values.The Chatbot asks for additonal information such as type of the food the user wants to eat.Based on the input from the user , the chatbot returns a response.


![gobot_database.png](https://github.com/deepmipt/DeepPavlov/blob/master/examples/img/gobot_database.png?raw=1)

figure source: (https://github.com/deepmipt/DeepPavlov/blob/master/examples/img/gobot_database.png?raw=1 )


          >> database([{'pricerange': 'costly', 'area': 'center'}])


          Out[1]: 
        [[{'name': 'Subway',
           'food': 'american',
           'pricerange': 'costly',
           'area': 'center',
           'addr': '2 metrotech,brooklyn',
           'phone': '01223 244277',
           'postcode': 'c.b 1, 7 d.y'},
          {'name': 'Taco bell',
           'food': 'mexican',
           'pricerange': 'costly',
           'area': 'center',
           'addr': '6 metrotech,brooklyn',
           'phone': '01223 327908',
           'postcode': 'c.b 1, 7 d.y'}]]
           

In [ ]:
!head -n 78 my_data/simple-dstc2-trn.json | tail +51

    {
      "speaker": 2,
      "text": "api_call area=\"south\" food=\"dontcare\" pricerange=\"cheap\"",
      "db_result": {
        "food": "chinese",
        "pricerange": "cheap",
        "area": "south",
        "addr": "cambridge leisure park clifton way cherry hinton",
        "phone": "01223 244277",
        "postcode": "c.b 1, 7 d.y",
        "name": "the lucky star"
      },
      "slots": [
        [
          "area",
          "south"
        ],
        [
          "pricerange",
          "cheap"
        ],
        [
          "food",
          "dontcare"
        ]
      ],
      "act": "api_call"
    },


We import Sqlite3Database package from DeepPavLov.we are creating a database of restaurents with primary key as 'name' of the restaurent and set the path of the database.

In [ ]:

from deeppavlov.core.data.sqlite_database import Sqlite3Database

database = Sqlite3Database(primary_keys=["name"],
                           save_path="my_bot/db.sqlite")

2020-05-17 04:20:15.505 WARNING in 'deeppavlov.core.models.serializable'['serializable'] at line 49: No load path is set for Sqlite3Database in 'infer' mode. Using save path instead
2020-05-17 04:20:15.515 INFO in 'deeppavlov.core.data.sqlite_database'['sqlite_database'] at line 66: Loading database from /content/my_bot/db.sqlite.


Feeding all the results of the api calls in the database.A table with keys is created -

            'food' : 'text'
            'area' : 'text'
            'addr' : 'text'
            'postcode': 'text'
            'pricerange' : 'text'
            'phone' : 'text'
            'name' : 'text'

  A total of     api calls are made in the dataset.

In [ ]:

db_results = []

for dialog in iterator.gen_batches(batch_size=1, data_type='all'):
    turns_x, turns_y = dialog
    db_results.extend(x['db_result'] for x in turns_x[0] if x.get('db_result'))

print(f"Adding {len(db_results)} items.")
if db_results:
    database.fit(db_results)

Adding 1780 items.


Retreiving restaurents corresponding to attribute values 'pricerange : 'cheap' and 'area': 'south'
we have reaturents with matching attribute values.

In [ ]:
database([{'pricerange': 'cheap', 'area': 'south'}])

[[{'addr': 'cambridge leisure park clifton way',
   'area': 'south',
   'food': 'portuguese',
   'name': 'nandos',
   'phone': '01223 327908',
   'postcode': 'c.b 1, 7 d.y',
   'pricerange': 'cheap'},
  {'addr': 'cambridge leisure park clifton way cherry hinton',
   'area': 'south',
   'food': 'chinese',
   'name': 'the lucky star',
   'phone': '01223 244277',
   'postcode': 'c.b 1, 7 d.y',
   'pricerange': 'cheap'}]]

In [ ]:
!ls my_bot

checkpoint		   model.index	slotfill
db.sqlite		   model.json	slotfill_config.json
model.data-00000-of-00001  model.meta	word.dict


## 2. Build Slot Filler

Solt filler is used to find the slot values in the user input.
Ex: " Suggest me cheap Indian restaurent"

slot values :
                  'pricerange' : 'cheap'
                  'food' : 'Indian'

sllot filler facilitates in finding the corresponding slot values in the user input.

slots_vals.json defines the slot type and slot values for the following dataset.

we need to provide all slot type.Ex; 'pricerange','food','name' etc
we need to provide all possible values of the slot type.
for example the user may input the type of food differently.Ex: He may input India instead of indian.Therefore we have to provide all the possible values of the slot type.


In [ ]:
from deeppavlov.download import download_decompress

download_decompress(url='http://files.deeppavlov.ai/deeppavlov_data/dstc_slot_vals.tar.gz',
                    download_path='my_bot/slotfill')

2020-05-17 04:20:17.244 INFO in 'deeppavlov.core.data.utils'['utils'] at line 80: Downloading from http://files.deeppavlov.ai/deeppavlov_data/dstc_slot_vals.tar.gz to my_bot/slotfill/dstc_slot_vals.tar.gz
100%|██████████| 1.62k/1.62k [00:00<00:00, 2.54MB/s]
2020-05-17 04:20:17.249 INFO in 'deeppavlov.core.data.utils'['utils'] at line 242: Extracting my_bot/slotfill/dstc_slot_vals.tar.gz archive into my_bot/slotfill


In [ ]:
!ls my_bot/slotfill

dstc_slot_vals.json


Some of the slot type with corresponding slot values

In [ ]:
!head -n 10 my_bot/slotfill/dstc_slot_vals.json

{
    "food": {
        "caribbean": [
            "carraibean",
            "carribean",
            "caribbean"
        ],
        "kosher": [
            "kosher"
        ],


In [ ]:
from deeppavlov import configs
from deeppavlov.core.common.file import read_json

slotfill_config = read_json(configs.ner.slotfill_simple_dstc2_raw)

In [ ]:
slotfill_config['metadata']['variables']['DATA_PATH'] = 'my_data'
slotfill_config['metadata']['variables']['SLOT_VALS_PATH'] = 'my_bot/slotfill/dstc_slot_vals.json'


Evaluating the model(slotfiller)

In [ ]:
from deeppavlov import evaluate_model

slotfill = evaluate_model(slotfill_config);

2020-05-17 04:20:19.643 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-trn.json]
2020-05-17 04:20:19.660 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-val.json]
2020-05-17 04:20:19.760 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-tst.json]
2020-05-17 04:20:19.851 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 282: There are 479 samples in train split.
2020-05-17 04:20:19.852 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 283: There are 6231 samples in valid split.
2020-05-17 04:20:19.856 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 284: There are 6345 samples in test split.


{"valid": {"eval_examples_count": 1444, "metrics": {"slots_accuracy": 0.9307}, "time_spent": "0:00:37"}}
{"test": {"eval_examples_count": 1386, "metrics": {"slots_accuracy": 0.9481}, "time_spent": "0:00:36"}}


The slotfiller was able to accurately identify user information in slots with an accuracy of 93% on validation set and 95% on test set.

Bulding the slotfiller using DeepPavLov config

In [ ]:
from deeppavlov import build_model

slotfill = build_model(slotfill_config)

Testing the slot filler

In [ ]:
slotfill(['i want cheap chinee food'])

[{'food': 'chinese', 'pricerange': 'cheap'}]

In [ ]:
slotfill(['I want India food'])

[{'food': 'indian'}]

Saving slotfill

In [ ]:
import json

json.dump(slotfill_config, open('my_bot/slotfill_config.json', 'wt'))

In [ ]:
!ls my_bot

checkpoint		   model.index	slotfill
db.sqlite		   model.json	slotfill_config.json
model.data-00000-of-00001  model.meta	word.dict


## 3. Build and Train a Bot

![gobot_policy.png](https://github.com/deepmipt/DeepPavlov/blob/master/examples/img/gobot_policy.png?raw=1)

 Source : (https://github.com/deepmipt/DeepPavlov/blob/master/examples/img/gobot_policy.png?raw=1)

 Based on the utterence and slot values the policy module in the dialogu manager decides which action to be taken.The policy in the bot is done by Recurrent neural networks with a dense layer of softmax function on top.Based on the response generated by the network,chatbot makes api calls to retrieve corresponding information.The ouputs from the database are used with the templates to provide the response to the user.

 The response generated by RNN and softmax are sift decisions.The action taken is decided on the corresponding probability of the action.

List of templates in the dataset

In [ ]:
!head -n 10 my_data/simple-dstc2-templates.txt

api_call	api_call area="#area" food="#food" pricerange="#pricerange"
bye	You are welcome!
canthear	Sorry, I can't hear you.
canthelp_area	I'm sorry but there is no #area american restaurant in the #area of town.
canthelp_area_food	Sorry there is no #food restaurant in the #area of town.
canthelp_area_food_pricerange	Sorry there is no #pricerange restaurant in the #area of town serving #food food.
canthelp_area_pricerange	Sorry there is no #pricerange restaurant in the #area of town serving #area american food.
canthelp_food	I am sorry but there is no #food restaurant that matches your request.
canthelp_food_pricerange	Sorry there is no #food restaurant in the #pricerange price range.
confirm-domain	You are looking for a restaurant is that right?


Sample of training set of DSTC-2

In [ ]:
!head -n 24 my_data/simple-dstc2-trn.json

[
  [
    {
      "speaker": 2,
      "text": "Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?",
      "slots": [],
      "act": "welcomemsg"
    },
    {
      "speaker": 1,
      "text": "cheap restaurant",
      "slots": [
        [
          "pricerange",
          "cheap"
        ]
      ]
    },
    {
      "speaker": 2,
      "text": "What kind of food would you like?",
      "slots": [],
      "act": "request_food"
    },


There are five main sections in DeepPavLov config 

* Dataset reader - It configures dataset reader component.Manages downloading data and saving it on the disk/drive.
* Dataset Iterator- It configures batch generator component.
* metadata - Contains extra information about data use and telegram bot parameters.
* train - It configures training process.Manages the size of the batch and number of training epochs.
* chainer - It specifies data flow and the order in which the components will be run.





In [ ]:
from deeppavlov import configs
from deeppavlov.core.common.file import read_json

gobot_config = read_json(configs.go_bot.gobot_simple_dstc2)

In [ ]:
gobot_config['chainer']['pipe'][-1]['embedder'] = None

In [ ]:
gobot_config['chainer']['pipe'][-1]['database'] = {
    'class_name': 'sqlite_database',
    'primary_keys': ["name"],
    'save_path': 'my_bot/db.sqlite'
}

### Levenshtein Distance

"Levenshtein distance between two words is the minimum number of single-character edits (i.e. insertions, deletions, or substitutions) required to change one word into the other." [link text](https://dzone.com/articles/the-levenshtein-algorithm-1).

For ex, 
take two words:   


                    local    &    global
                    Substituting 'c' for 'a'
                    inserting  'g' at front

        Levenshtein distance between the words is 2.


Configure bot to use levenshtein distance based slot filler:


In [ ]:
gobot_config['chainer']['pipe'][-1]['slot_filler']['config_path'] = 'my_bot/slotfill_config.json'

Slot types 

In [ ]:
gobot_config['chainer']['pipe'][-1]['tracker']['slot_names'] = ['pricerange', 'this', 'area', 'food']

Configuring the bot to use saved templates :

In [ ]:
gobot_config['chainer']['pipe'][-1]['template_type'] = 'DefaultTemplate'
gobot_config['chainer']['pipe'][-1]['template_path'] = 'my_data/simple-dstc2-templates.txt'

In [ ]:
gobot_config['metadata']['variables']['DATA_PATH'] = 'my_data'
gobot_config['metadata']['variables']['MODEL_PATH'] = 'my_bot'

Traning the network 

Batch size is 8.Therefore every bactch contains 8 dialogues from the user and 8 dialogues from the chatbot as training data.
Maximum number of batches is 100(Original notebook had 250 but due to excessive time for training I have taken only 100.There is no significant drop in accuracy i.e, 0.48 for 100 batches and 0.5 for 250 batches)


In [ ]:
from deeppavlov import train_model

gobot_config['train']['batch_size'] = 8 # batch size
gobot_config['train']['max_batches'] = 100 # maximum number of training batches
gobot_config['train']['log_on_k_batches'] = 20
gobot_config['train']['val_every_n_batches'] = 40 # evaluate on full 'valid' split each n batches
gobot_config['train']['log_every_n_batches'] = 40 # evaluate on 20 batches of 'train' split every n batches

train_model(gobot_config);

2020-05-17 04:21:35.601 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-trn.json]
2020-05-17 04:21:35.612 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-val.json]
2020-05-17 04:21:36.72 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-tst.json]
2020-05-17 04:21:36.145 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 282: There are 479 samples in train split.
2020-05-17 04:21:36.146 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 283: There are 6231 samples in valid split.
2020-05-17 04:21:36.147 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 284: There are 6345 samples in test split.
2020-05-17 04:21:36.164 INFO in 'deeppavlov.core.data.simple_vocab'['simple_vocab'

INFO:tensorflow:Restoring parameters from /content/my_bot/model


2020-05-17 04:23:28.46 INFO in 'deeppavlov.core.trainers.nn_trainer'['nn_trainer'] at line 198: Initial best per_item_dialog_accuracy of 0.4807


{"valid": {"eval_examples_count": 575, "metrics": {"per_item_dialog_accuracy": 0.4807}, "time_spent": "0:01:50", "epochs_done": 0, "batches_seen": 0, "train_examples_seen": 0, "impatience": 0, "patience_limit": 10}}
{"train": {"eval_examples_count": 50, "metrics": {"per_item_dialog_accuracy": 0.8309}, "time_spent": "0:02:46", "epochs_done": 5, "batches_seen": 40, "train_examples_seen": 290, "learning_rate": 0.003, "momentum": 0.95, "loss": 0.373494853079319}}


2020-05-17 04:26:12.658 INFO in 'deeppavlov.core.trainers.nn_trainer'['nn_trainer'] at line 206: Improved best per_item_dialog_accuracy of 0.4967
2020-05-17 04:26:12.660 INFO in 'deeppavlov.core.trainers.nn_trainer'['nn_trainer'] at line 208: Saving model
2020-05-17 04:26:12.661 INFO in 'deeppavlov.core.models.tf_model'['tf_model'] at line 75: [saving model to /content/my_bot/model]
2020-05-17 04:26:12.932 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 711: [saving parameters to /content/my_bot/model.json]


{"valid": {"eval_examples_count": 575, "metrics": {"per_item_dialog_accuracy": 0.4967}, "time_spent": "0:04:35", "epochs_done": 5, "batches_seen": 40, "train_examples_seen": 290, "impatience": 0, "patience_limit": 10}}
{"train": {"eval_examples_count": 50, "metrics": {"per_item_dialog_accuracy": 0.8685}, "time_spent": "0:05:30", "epochs_done": 11, "batches_seen": 80, "train_examples_seen": 574, "learning_rate": 0.003, "momentum": 0.95, "loss": 0.20555392149835824}}


2020-05-17 04:28:57.548 INFO in 'deeppavlov.core.trainers.nn_trainer'['nn_trainer'] at line 211: Did not improve on the per_item_dialog_accuracy of 0.4967


{"valid": {"eval_examples_count": 575, "metrics": {"per_item_dialog_accuracy": 0.4945}, "time_spent": "0:07:20", "epochs_done": 11, "batches_seen": 80, "train_examples_seen": 574, "impatience": 1, "patience_limit": 10}}


2020-05-17 04:29:20.829 INFO in 'deeppavlov.core.data.simple_vocab'['simple_vocab'] at line 115: [loading vocabulary from /content/my_bot/word.dict]
2020-05-17 04:29:20.832 WARNING in 'deeppavlov.core.models.serializable'['serializable'] at line 49: No load path is set for Sqlite3Database in 'infer' mode. Using save path instead
2020-05-17 04:29:20.836 INFO in 'deeppavlov.core.data.sqlite_database'['sqlite_database'] at line 66: Loading database from /content/my_bot/db.sqlite.
2020-05-17 04:29:21.378 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 164: [loading templates from /content/my_data/simple-dstc2-templates.txt]
2020-05-17 04:29:21.380 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 167: 46 templates loaded.
2020-05-17 04:29:21.382 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 215: Calculated input size for `GoalOrientedBotNetwork` is 205
2020-05-17 04:29:22.210 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 253: [initi

INFO:tensorflow:Restoring parameters from /content/my_bot/model
{"valid": {"eval_examples_count": 575, "metrics": {"per_item_dialog_accuracy": 0.4967}, "time_spent": "0:01:49"}}


2020-05-17 04:32:58.584 INFO in 'deeppavlov.core.data.simple_vocab'['simple_vocab'] at line 115: [loading vocabulary from /content/my_bot/word.dict]
2020-05-17 04:32:58.587 WARNING in 'deeppavlov.core.models.serializable'['serializable'] at line 49: No load path is set for Sqlite3Database in 'infer' mode. Using save path instead
2020-05-17 04:32:58.592 INFO in 'deeppavlov.core.data.sqlite_database'['sqlite_database'] at line 66: Loading database from /content/my_bot/db.sqlite.


{"test": {"eval_examples_count": 576, "metrics": {"per_item_dialog_accuracy": 0.4883}, "time_spent": "0:01:48"}}


2020-05-17 04:32:59.90 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 164: [loading templates from /content/my_data/simple-dstc2-templates.txt]
2020-05-17 04:32:59.92 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 167: 46 templates loaded.
2020-05-17 04:32:59.93 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 215: Calculated input size for `GoalOrientedBotNetwork` is 205
2020-05-17 04:32:59.896 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 253: [initializing `GoalOrientedBot` from saved]
2020-05-17 04:32:59.897 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 717: [loading parameters from /content/my_bot/model.json]
2020-05-17 04:32:59.904 INFO in 'deeppavlov.core.models.tf_model'['tf_model'] at line 51: [loading model from /content/my_bot/model]


INFO:tensorflow:Restoring parameters from /content/my_bot/model


Calculating accuracy of the model

In [ ]:
from deeppavlov import evaluate_model

evaluate_model(gobot_config);

2020-05-17 04:32:59.981 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-trn.json]
2020-05-17 04:32:59.994 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-val.json]
2020-05-17 04:33:00.476 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 290: [loading dialogs from /content/my_data/simple-dstc2-tst.json]
2020-05-17 04:33:00.542 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 282: There are 479 samples in train split.
2020-05-17 04:33:00.543 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 283: There are 6231 samples in valid split.
2020-05-17 04:33:00.544 INFO in 'deeppavlov.dataset_readers.dstc2_reader'['dstc2_reader'] at line 284: There are 6345 samples in test split.
2020-05-17 04:33:00.562 INFO in 'deeppavlov.core.data.simple_vocab'['simple_vocab

INFO:tensorflow:Restoring parameters from /content/my_bot/model
{"valid": {"eval_examples_count": 575, "metrics": {"per_item_dialog_accuracy": 0.4967}, "time_spent": "0:01:49"}}
{"test": {"eval_examples_count": 576, "metrics": {"per_item_dialog_accuracy": 0.4883}, "time_spent": "0:01:47"}}


## 4. Interact with Bot

Intializing the bot

In [ ]:
from deeppavlov import build_model

bot = build_model(gobot_config)

2020-05-17 04:36:37.615 INFO in 'deeppavlov.core.data.simple_vocab'['simple_vocab'] at line 115: [loading vocabulary from /content/my_bot/word.dict]
2020-05-17 04:36:37.618 WARNING in 'deeppavlov.core.models.serializable'['serializable'] at line 49: No load path is set for Sqlite3Database in 'infer' mode. Using save path instead
2020-05-17 04:36:37.621 INFO in 'deeppavlov.core.data.sqlite_database'['sqlite_database'] at line 66: Loading database from /content/my_bot/db.sqlite.
2020-05-17 04:36:38.167 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 164: [loading templates from /content/my_data/simple-dstc2-templates.txt]
2020-05-17 04:36:38.169 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 167: 46 templates loaded.
2020-05-17 04:36:38.170 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 215: Calculated input size for `GoalOrientedBotNetwork` is 205
2020-05-17 04:36:39.49 INFO in 'deeppavlov.models.go_bot.network'['network'] at line 253: [initia

INFO:tensorflow:Restoring parameters from /content/my_bot/model


Interacting with the bot

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['Indian food'])

['What kind of food would you like?']

In [ ]:
bot(['give me the address'])

2020-05-17 04:36:39.414 INFO in 'deeppavlov.models.go_bot.tracker'['tracker'] at line 210: Made api_call with {'food': 'indian'}, got 22 results.


['The nirala serves indian food.']

In [ ]:
bot(['Give me the phone number'])

['The phone number of the nirala is 01223 360966.']

### EXTENSION

In this example the user waqnts to eat a burger.The response from the chatbot is vague and does not relate with the input of the user as it cannot find the slot value for 'food'.As there is no value of any slot_type corresponding to any of the words in the example the chatbot ouputs a default value to the user. 

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['i want to eat a burger'])

['What kind of food would you like?']

In [ ]:
bot(['i want to eat a chicken burger'])

['What kind of food would you like?']

In this example, two type of food types are given as input but the chatbot only has the first value in the order of utterence as the value of 'food'.It does not take into consideration indian and suggests restaurents based on slot_type 'food' : chinese

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['chinese and Indain food'])

['What kind of food would you like?']

In this example, the user wants to eat vegan food.But the database does not contain 'vegetarian' as any of its slot values.Therefore the chatbot returns an empty template.

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['I want to eat vegetarian food'])

['What kind of food would you like?']

In [ ]:
bot(['I want to eat non-vegetarian food'])

2020-05-17 04:36:39.880 INFO in 'deeppavlov.models.go_bot.tracker'['tracker'] at line 210: Made api_call with {'food': 'vegetarian'}, got 0 results.


['I am sorry but there is no vegetarian restaurant that matches your request.']

In this example, the user asks if the restaurent has vegan options.The chatbot is picks up the words options and provides the address of the restaurent rather than answering the question of the user.

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['I want to eat in a costly restaurent'])

['What kind of food would you like?']

In [ ]:
bot(['I like to have meaxican food'])

2020-05-17 04:36:40.56 INFO in 'deeppavlov.models.go_bot.tracker'['tracker'] at line 210: Made api_call with {'food': 'mexican'}, got 1 results.


['Chiquito restaurant bar serves mexican food.']

In [ ]:
bot(['Does it have vegan options'])

['You are looking for a restaurant is that right?']

In this example, the user directly asks the address of the restaurent.The NLU picks up the word address but it does not have the values of other attributes like ' pricerange' , 'food', 'area' therefore it returns an empty template to the user.

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['Can you get the address of chiquito restaurent'])

['Sure, #name is on #addr.']

In [ ]:
bot(['i want their phone number too'])

['The phone number of #name is #phone.']

In this example, The user wants to eat desserts.As there is no value corresponding to any slot values in the user utterance,the chatbot by default asks the user about the type of food.

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['I like desserts'])

['What kind of food would you like?']

In [ ]:
bot.reset()

In this example, the chatbot performs well with identifying all three slot_values but when the user requests other options the chatbot gives the user address for the restaurent suggested rather than providing him other options.

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['I want to eat Indian food in the center of the town.I prefer cheap options.'])

2020-05-17 04:36:40.506 INFO in 'deeppavlov.models.go_bot.tracker'['tracker'] at line 210: Made api_call with {'food': 'indian', 'pricerange': 'cheap', 'area': 'centre'}, got 3 results.


['Mahal of cambridge serves indian food.']

In [ ]:
bot(['Any other options?'])

['You are looking for a restaurant is that right?']

In this example,the user changes his decision on the area,but the chatbot's NLU misinterprets it and outputs the address of the restaurent suggested earlier. 

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['I would like to go restaurent in north of the town'])

2020-05-17 04:36:40.692 INFO in 'deeppavlov.models.go_bot.tracker'['tracker'] at line 210: Made api_call with {'area': 'north'}, got 11 results.


['Restaurant two two serves french food.']

In [ ]:
bot(['ohh then suggest me restaurents in south then'])

['Sure, restaurant two two is on 22 chesterton road chesterton.']

In this example the user mentions two pricerange,but the chatbot only picks up the first realted slot_value from the utterence leaving the second slot_value. This results in the chatbot suggesting only cheap options rather than suggesting both cheap and moderate options.

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['I can afford cheap and moderate food'])

['What part of town do you have in mind?']

In [ ]:
bot(['indian'])

2020-05-17 04:36:40.899 INFO in 'deeppavlov.models.go_bot.tracker'['tracker'] at line 210: Made api_call with {'pricerange': 'cheap', 'food': 'indian'}, got 4 results.


['Royal spice serves indian food.']

In this example,the chatbot cannot determine the slot_value for any of the slot_types.Therefore it outputs the default response.

In [ ]:
bot.reset()

In [ ]:
bot(['hi, i want to eat, can you suggest a place to go?'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['Continental cuisine'])

['What part of town do you have in mind?']

In [ ]:
bot(['south would be great'])

2020-05-17 04:36:41.38 INFO in 'deeppavlov.models.go_bot.tracker'['tracker'] at line 210: Made api_call with {'area': 'south'}, got 9 results.


['Pizza hut cherry hinton is a nice place in the south of town serving tasty italian food.']

In [ ]:
bot.reset()

In this example, the chatbot could not identify the slot_value for any of the attributes.

In [ ]:
bot(['Hi'])

['Hello, welcome to the Cambridge restaurant system. You can ask for restaurants by area, price range or food type. How may I help you?']

In [ ]:
bot(['bye'])

['What part of town do you have in mind?']

In [ ]:
bot(['bye'])

['What part of town do you have in mind?']